stdlib/public/core/Sort.swift

```diff
@@ -257,10 +257,42 @@ extension MutableCollection where Self: RandomAccessCollection {
   }
 }
 
+extension Collection where Self: BidirectionalCollection {
+  /// This is a simple binary search like algorithm which is finding an index
+  /// where the first element that is greater than `key` should be located.
+  /// This algorithm works only on sorted collection.
+  @inlinable
+  internal func _findUpperBound(
+    forKey key: Element,
+    within range: Range<Index>,
+    by areInIncreasingOreder: (Element, Element) throws -> Bool
+    ) rethrows -> Index {
+    
+    var start = range.lowerBound
+    var end   = range.upperBound
+    
+    while start < end {
+      let mid = index(start,
+                      offsetBy: distance(from: start, to: end) &>> 1)
```

Comment:
nit: I think this would fit within 80 col on one line? unless the intention is clarity, in which case I'd just create an intermediate `half` variable.